File: src/com/yhsif/onepwd/HmacMd5.java
package com.yhsif.onepwd;

import static java.nio.charset.StandardCharsets.UTF_8;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/** 
 * An HMAC-MD5 implementation.
 */
public class HmacMd5 {

	public static final int BLOCK_SIZE = 64;
	public static final int HASH_SIZE = 16;
	public static final byte O_KEY_PAD = 0x5c;
	public static final byte I_KEY_PAD = 0x36;

	public static byte[] md5(byte[] message) throws NoSuchAlgorithmException {
		MessageDigest digester = MessageDigest.getInstance("MD5");
		digester.update(message);
		return digester.digest();
	}

	public static byte[] hmac(String key, String message) throws NoSuchAlgorithmException {
		byte keys[] = new byte[BLOCK_SIZE];
		if (key.length() > BLOCK_SIZE) {
			System.arraycopy(md5(key.getBytes(UTF_8)), 0, keys, 0, HASH_SIZE);
		} else {
			System.arraycopy(key.getBytes(UTF_8), 0, keys, 0, key.length());
		}
		byte oKeyBuf[] = new byte[BLOCK_SIZE + HASH_SIZE];
		byte iKeyBuf[] = new byte[BLOCK_SIZE + message.length()];
		for (int i = 0; i < BLOCK_SIZE; i++) {
			oKeyBuf[i] = (byte) (O_KEY_PAD ^ keys[i]);
			iKeyBuf[i] = (byte) (I_KEY_PAD ^ keys[i]);
		}
		System.arraycopy(message.getBytes(UTF_8), 0, iKeyBuf, BLOCK_SIZE, message.length());
		System.arraycopy(md5(iKeyBuf), 0, oKeyBuf, BLOCK_SIZE, HASH_SIZE);
		return md5(oKeyBuf);
	}

}

File: src/com/yhsif/onepwd/HmacMd5.kt
package com.yhsif.onepwd

import java.security.MessageDigest

/**
 * An HMAC-MD5 implementation.
 */
object HmacMd5 {
    const val BLOCK_SIZE = 64
    const val HASH_SIZE = 16
    const val O_KEY_PAD: Byte = 0x5c
    const val I_KEY_PAD: Byte = 0x36

    fun md5(message: ByteArray): ByteArray {
        MessageDigest.getInstance("MD5").let { digester ->
            digester.update(message)
            return digester.digest()
        }
    }

    fun hmac(key: String, message: String): ByteArray {
        val keys = ByteArray(BLOCK_SIZE)
        if (key.length > BLOCK_SIZE) {
            System.arraycopy(md5(key.toByteArray()), 0, keys, 0, HASH_SIZE)
        } else {
            System.arraycopy(key.toByteArray(), 0, keys, 0, key.length)
        }
        val oKeyBuf = ByteArray(BLOCK_SIZE + HASH_SIZE)
        val iKeyBuf = ByteArray(BLOCK_SIZE + message.length)
        for (i in 0 until BLOCK_SIZE) {
            oKeyBuf[i] = (O_KEY_PAD.toInt() xor keys[i].toInt()).toByte()
            iKeyBuf[i] = (I_KEY_PAD.toInt() xor keys[i].toInt()).toByte()
        }
        System.arraycopy(
            message.toByteArray(),
            0,
            iKeyBuf,
            BLOCK_SIZE,
            message.length
        )
        System.arraycopy(md5(iKeyBuf), 0, oKeyBuf, BLOCK_SIZE, HASH_SIZE)
        return md5(oKeyBuf)
    }
}
